File: app/src/main/java/com/bigbang/googlebooksfirebase/database/FirebaseEvents.java
package com.bigbang.googlebooksfirebase.database;

import com.bigbang.googlebooksfirebase.model.Book;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.annotations.NonNull;
import io.reactivex.subjects.PublishSubject;

import static com.bigbang.googlebooksfirebase.util.DebugLogger.logDebug;

public class FirebaseEvents {

    private DatabaseReference booksReference;
    private PublishSubject<List<Book>> bookListObservable = PublishSubject.create();

    public FirebaseEvents() {
        this.booksReference = FirebaseDatabase.getInstance().getReference().child("books");
    }

    public void addNewBook(Book book){
        String bookKey = booksReference.push().getKey();
        if (bookKey != null)
            booksReference.child(bookKey).setValue(book);
        else
            logDebug("db update failed");
    }

    public Observable<List<Book>> getBookList(){
        List<Book> bookList = new ArrayList<>();

        booksReference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                logDebug("inside FirebaseHelper -> onDataChange()");

                for(DataSnapshot currentSnap : dataSnapshot.getChildren()){
                    Book currentBook = currentSnap.getValue(Book.class);
                    bookList.add(currentBook);
                }
                bookListObservable.onNext(bookList);
            }
            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
                logDebug("db error");
            }
        });
        logDebug("Book List Size before return "+ bookList.size());
        return bookListObservable;
    }

}

File: app/src/main/java/com/bigbang/googlebooksfirebase/database/FirebaseEvents.kt
package com.bigbang.googlebooksfirebase.database

import com.bigbang.googlebooksfirebase.model.Book
import com.bigbang.googlebooksfirebase.util.DebugLogger.logDebug
import com.google.firebase.database.*
import io.reactivex.Observable
import io.reactivex.subjects.PublishSubject
import java.util.*

class FirebaseEvents {
    private val booksReference: DatabaseReference
    private val bookListObservable = PublishSubject.create<List<Book>>()
    fun addNewBook(book: Book?) {
        val bookKey = booksReference.push().key
        if (bookKey != null) booksReference.child(bookKey).setValue(book) else logDebug("db update failed")
    }

    //logDebug("Book List Size before return " + bookList.size());
    val bookList: Observable<List<Book>>
        get() {
            val bookList: MutableList<Book> = ArrayList()
            booksReference.addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    logDebug("inside FirebaseHelper -> onDataChange()")
                    for (currentSnap in dataSnapshot.children) {
                        val currentBook = currentSnap.getValue(Book::class.java)!!
                        bookList.add(currentBook)
                    }
                    bookListObservable.onNext(bookList)
                }

                override fun onCancelled(databaseError: DatabaseError) {
                    logDebug("db error")
                }
            })
            //logDebug("Book List Size before return " + bookList.size());
            return bookListObservable
        }

    init {
        booksReference = FirebaseDatabase.getInstance().reference.child("books")
    }
}